File: src/main/java/org/jcodec/containers/raw/RawMuxer.java
package org.jcodec.containers.raw;

import java.io.IOException;

import org.jcodec.common.AudioCodecMeta;
import org.jcodec.common.Muxer;
import org.jcodec.common.MuxerTrack;
import org.jcodec.common.VideoCodecMeta;
import org.jcodec.common.io.SeekableByteChannel;
import org.jcodec.common.model.Packet;

public class RawMuxer implements Muxer, MuxerTrack {

    private SeekableByteChannel ch;
    private boolean hasVideo;
    private boolean hasAudio;

    public RawMuxer(SeekableByteChannel destStream) {
        this.ch = destStream;
    }

    @Override
    public MuxerTrack addVideoTrack(VideoCodecMeta meta) {
        if (hasAudio)
            throw new RuntimeException("Raw muxer supports either video or audio track but not both.");
        hasVideo = true;
        return this;
    }

    @Override
    public MuxerTrack addAudioTrack(AudioCodecMeta meta) {
        if (hasVideo)
            throw new RuntimeException("Raw muxer supports either video or audio track but not both.");
        hasAudio = true;
        return this;
    }

    @Override
    public void finish() throws IOException {
    }

    @Override
    public void addFrame(Packet outPacket) throws IOException {
        ch.write(outPacket.getData().duplicate());
    }
}

File: src/main/java/org/jcodec/containers/raw/RawMuxer.kt
package org.jcodec.containers.raw

import org.jcodec.common.*
import org.jcodec.common.io.SeekableByteChannel
import org.jcodec.common.model.Packet
import java.io.IOException

class RawMuxer(private val ch: SeekableByteChannel) : Muxer, MuxerTrack {
    private var hasVideo = false
    private var hasAudio = false
    override fun addVideoTrack(codec: Codec, meta: VideoCodecMeta): MuxerTrack {
        if (hasAudio) throw RuntimeException("Raw muxer supports either video or audio track but not both.")
        hasVideo = true
        return this
    }

    override fun addAudioTrack(codec: Codec, meta: AudioCodecMeta): MuxerTrack {
        if (hasVideo) throw RuntimeException("Raw muxer supports either video or audio track but not both.")
        hasAudio = true
        return this
    }

    @Throws(IOException::class)
    override fun finish() {
    }

    @Throws(IOException::class)
    override fun addFrame(outPacket: Packet) {
        ch.write(outPacket.getData().duplicate())
    }

}